airflow/ti_deps/deps/trigger_rule_dep.py

```diff
@@ -381,6 +381,8 @@ def _evaluate_direct_relatives() -> Iterator[TIDepStatus]:
                 elif trigger_rule == TR.ALL_SKIPPED:
                     if success or failed:
                         new_state = TaskInstanceState.SKIPPED
+                    elif upstream_done and upstream_failed:
+                        new_state = TaskInstanceState.UPSTREAM_FAILED
```

Comment:
So....

Why should upstream failed not be treated the exact same as failed?  

I agree that it's weird that this rule results in "skip" when upstream is "fail".  But, that's how it is.  It seems that upstream_failed should treated no different from failed.

Reply:
That is a good question I feel I am not qualified to answer 😅 
I see the argument for both, fail and upstream fail causing the same response (that is what I did at first)
and also for upstream fail to be different because who knows if it could have been skipped like @hussein-awala suggested 🤔

Reply:
> Why should upstream failed not be treated the exact same as failed?

For me, when an upstream TI ends with a success or failed state, we know that our TI cannot be run our TI, so we can skip (or fast skip) it.
However, when there is no success/failed upstream and we have an upstream with upstream_failed state, we cannot skip the TI because we don't know the state of this upstream when we run it, so by switching to `upstream_failed` we tell the user "Sorry we cannot determine the state of your TI".

But if you just consider `upstream_failed` equivalent to `failed` in this situation, TBH I don't have a strong opinion on that.

Reply:
Yeah I chatted about this with some folks at my company... It does seem that the right way to think about `upstream_failed` as that it's effectively a subtype of failed.  Essentially, it is failed _with reason_ "upstream failed".  

And, for evidence of this, we can look at how other trigger rules work.

For example with "all failed", the rule is `All upstream tasks are in a failed or upstream_failed state`.
So for that rule, clearly upstream failed is treated same as failed.

I think it should be the same with all_skipped.

Reply:
I was watching the discussion and scratched my head a bit and also have no clear opinion.
I feel similar like the state model in https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/tasks.html#task-instances we might need to add a mapping table to how a task responds depending on trigger rule and upstream dependencies.

Reply:
cc @lzdanski :)

Reply:
> I was watching the discussion and scratched my head a bit and also have no clear opinion. I feel similar like the state model in https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/tasks.html#task-instances we might need to add a mapping table to how a task responds depending on trigger rule and upstream dependencies.

Yeah, the question of whether the task runs or not I think is pretty clearly documented in trigger rules docs.  But the question of, when it doesn't run, what will the end state be, don't think that is well documented if at all, and could use some attention.  Sometimes it's subject to a race condition -- e.g. depends what the state of the first non-success is.  Anyway, I think the fix in this PR seems good to me.